### PR TITLE
New Clippy source test: lint-attributes-organization

### PR DESCRIPTION
**Pull request by samueltardieu**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/17050)*

Add a new Clippy source test `lint-attributes-organization` which checks that:

- top-level inner lint-related attributes (`allow`/`deny`/`expect`/`forbid`/`warn`) are not repeated in tests
- unprefixed lint names (`unused`) come before prefixed ones (`clippy::cast_precision_loss`)
- lint names are ordered in each category

changelog: none

### CODE REVIEW

**Comment by samueltardieu** [comment]
Since we talked about this yesterday at the AllHands:
r? ada4a

File: tests/ui/arithmetic_side_effects.rs [hunk 1]
```diff
@@ -2,16 +2,12 @@
...
-    clippy::unnecessary_owned_empty_strings,
-    arithmetic_overflow,
-    unconditional_panic
-)]
+#![expect(clippy::unnecessary_owned_empty_strings)]
+#![expect(clippy::op_ref)]
+#![expect(clippy::no_effect)]
+#![expect(clippy::identity_op)]
+#![expect(clippy::erasing_op)]
+#![expect(clippy::assign_op_pattern)]
```
Comment (ada4a) [inline tests/ui/arithmetic_side_effects.rs]: having these all separated is kind of obnoxious, could you please do another pass which combines them? also not sure why the order has been reversed, but that's fine...

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/423f9652098ae0761bec00be5e2d04c78a879f08..6a5bcbba9f45031da13591afff0e7c943af329e9)*

File: tests/ui/assertions_on_constants.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#![allow(non_fmt_panics, clippy::needless_bool, clippy::eq_op)]
+#![expect(clippy::eq_op)]
 #![expect(clippy::manual_assert_eq)]
```
Comment (ada4a) [inline tests/ui/assertions_on_constants.rs]: nit: these would be good to combine too..

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/423f9652098ae0761bec00be5e2d04c78a879f08..6a5bcbba9f45031da13591afff0e7c943af329e9)*

File: tests/ui/assigning_clones.rs [hunk 3]
```diff
@@ -1,9 +1,8 @@
-#![allow(unused)]
-#![allow(clippy::redundant_clone)]
+#![expect(clippy::redundant_clone)]
 #![allow(clippy::ptr_arg)] // https://github.com/rust-lang/rust-clippy/issues/10612
```
Comment (ada4a) [inline tests/ui/assigning_clones.rs]: the linked issue has since been fixed, and it looks like removing this line doesn't break anything. could you please remove it?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/423f9652098ae0761bec00be5e2d04c78a879f08..6a5bcbba9f45031da13591afff0e7c943af329e9)*

**Review by ada4a — CHANGES_REQUESTED** [review]

File: tests/ui/arithmetic_side_effects.rs (hunk 1 above)
Comment (samueltardieu) [reply tests/ui/arithmetic_side_effects.rs]: Done. I've also reordered all attribute names.

File: tests/ui-cargo/multiple_inherent_impl/crate_fail/src/main.rs [hunk 4]
```diff
@@ -1,4 +1,3 @@
-#![allow(dead_code)]
```
Comment (ada4a) [inline tests/ui-cargo/multiple_inherent_impl/crate_fail/src/main.rs:1]: removing this causes a bunch of new warnings in [`tests/ui-cargo/multiple_inherent_impl/crate_fail/Cargo.stderr`](https://github.com/rust-lang/rust-clippy/pull/17050/files#diff-25e35dc23db3ca534ba9f885d8b9320c8c3ee61a6d510261d6fb41f3c8c291cb), so we should probably bring it back...

File: tests/ui-cargo/multiple_inherent_impl/file_fail/src/main.rs [hunk 5]
```diff
@@ -1,4 +1,3 @@
-#![allow(dead_code)]
```
Comment (ada4a) [inline tests/ui-cargo/multiple_inherent_impl/file_fail/src/main.rs:1]: ditto

**Review by ada4a — CHANGES_REQUESTED** [review]
Apologies for only giving the review piecemeal, it's a bit too big to get through in one sitting (it also makes GitHub very laggy...)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/8a61907ac8ddde3cff0f6879b22a91975ae924b0..37d66426bf74e902d952c74ead1e6f54f3002c45)*

File: tests/ui-cargo/multiple_inherent_impl/crate_fail/src/main.rs (hunk 4 above)
Comment (samueltardieu) [reply tests/ui-cargo/multiple_inherent_impl/crate_fail/src/main.rs:1]: You're right, that's better. Done.

File: tests/ui-cargo/multiple_inherent_impl/file_fail/src/main.rs (hunk 5 above)
Comment (samueltardieu) [reply tests/ui-cargo/multiple_inherent_impl/file_fail/src/main.rs:1]: Done

**Comment by samueltardieu** [comment]
CI seems stuck, will rebase to repush.

**Comment by samueltardieu** [comment]
I've added two separate commits:
- do the same reorganization for `#![warn]` in tests
- add a new test that checks our test files and ensures that we don't duplicate top-level attributes, and that names are ordered within them

The new test is disabled from inside the compiler test suite as people fixing Clippy test suite as part of unrelated changes in the compiler, as is customary.

File: tests/lint-attributes-organization.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        }
+    }
+    diags
+}
+
+#[test]
+fn internal_sanity_check() {
+    let attribute_regex = attribute_regex();
+    check_errors(
+        "#![allow(\n  clippy::def, // Comment \n  clippy::abc\n)]\n#![allow(clippy::zyx, unknown)]",
```
Comment (ada4a) [inline tests/lint-attributes-organization.rs]: nit: splitting this over lines would make it more clear what's going on imo?

File: tests/ui-cargo/multiple_inherent_impl/module_fail/src/main.rs [hunk 7]
```diff
@@ -1,4 +1,3 @@
-#![allow(dead_code)]
```
Comment (ada4a) [inline tests/ui-cargo/multiple_inherent_impl/module_fail/src/main.rs:1]: this one should be retained as well

File: tests/ui/assigning_clones.rs [hunk 8]
```diff
@@ -1,9 +1,5 @@
-#![allow(unused)]
-#![allow(clippy::redundant_clone)]
-#![allow(clippy::ptr_arg)] // https://github.com/rust-lang/rust-clippy/issues/10612
-#![allow(clippy::needless_late_init)]
-#![allow(clippy::box_collection)]
-#![allow(clippy::boxed_local)]
+#![expect(clippy::boxed_local, clippy::needless_late_init)]
+#![allow(clippy::box_collection, clippy::ptr_arg)] // https://github.com/rust-lang/rust-clippy/issues/10612
```
Comment (ada4a) [inline tests/ui/assigning_clones.rs]: looks like `ptr_arg` snuck back in
```suggestion
#![allow(clippy::box_collection)]
```

**Review by ada4a — CHANGES_REQUESTED** [review]
There are still some subtle errors remaining, which makes me wary of merging this whole thing without going through it in earnest. Maybe we should do it in batches of ~100 files instead? Would also make rebasing less of a problem

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/ba8b78b0a89a83d227540770a4274c2840a046aa..c2a7808e235d0d938e074f2cce9787fa82cdc8f4)*

**Comment by samueltardieu** [comment]
> There are still some subtle errors remaining, which makes me wary of merging this whole thing without going through it in earnest. Maybe we should do it in batches of ~100 files instead? Would also make rebasing less of a problem
> 
> _[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/ba8b78b0a89a83d227540770a4274c2840a046aa..c2a7808e235d0d938e074f2cce9787fa82cdc8f4)_

What do you mean? You would prefer to have 20 pull requests? I can split it (01/20, 02/20, …, 20/20) and generate PRs. Or one PR with 20 commits?

File: tests/lint-attributes-organization.rs (hunk 6 above)
Comment (samueltardieu) [reply tests/lint-attributes-organization.rs]: Done

**Comment by ada4a** [comment]
> You would prefer to have 20 pull requests?

Well, not when you put it like that 😄 But yes, that's what I meant

**Comment by samueltardieu** [comment]
> > You would prefer to have 20 pull requests?
> 
> Well, not when you put it like that 😄 But yes, that's what I meant

I'm ok in splitting it, I just want to make sure you'll be ok with this afflux of (easier to review) PRs. My goal is to have them all merged ultimately, otherwise this is not worth doing it.

**Comment by samueltardieu** [comment]
I'll change this PR into a draft for the time being. The draft will contain all unmerged cleanup PRs (20 of them to start with), and on top of that the new test to ensure that everything is clean. When all cleanup PRs are merged, I'll un-draft this one to add the new test.

**Comment by ada4a** [comment]
> My goal is to have them all merged ultimately

That is my goal as well:)

> I just want to make sure you'll be ok with this afflux of (easier to review) PRs

Well that should definitely be easier to do than one big PR. Though you might consider only preparing 1 or 2 PRs in advance, to avoid rebase churn. What I want to say is that just splitting this PR into 20 parts and reviewing them one-by-one will probably not work: by the time we get to the 10th sub-PR, many tests will've changed under our feet, and so you'd need to run the cleanup on those changed versions anew.

**Comment by samueltardieu** [comment]
> Though you might consider only preparing 1 or 2 PRs in advance, to avoid rebase churn. 

The PRs are independent of each other (and don't conflict with each other as they have independent set of files), and all based on master. They should rarely require rebasing, and if that happens, only one PR will be concerned at a time.

**Comment by ada4a** [comment]
Sure, okay. Let's try it that way then

**Comment by samueltardieu** [comment]
r? @ada4a 
(since it is the logical consequence of the /20 series)

File: src/driver.rs [hunk 9]
```diff
@@ -1,8 +1,7 @@
 #![feature(rustc_private)]
 // warn on lints, that are included in `rust-lang/rust`s bootstrap
-#![warn(rust_2018_idioms, unused_lifetimes)]
-// warn on rustc internal lints
-#![warn(rustc::internal)]
+// and on rustc internal lints
+#![warn(rust_2018_idioms, unused_lifetimes, rustc::internal)]
```
Comment (ada4a) [inline src/driver.rs]: How about something like this, to preserve comment-to-lint correspondence?
```rs
#![warn(
    // warn on lints that are included in `rust-lang/rust`s bootstrap
    rust_2018_idioms,
    unused_lifetimes,
    // and on rustc internal lints
    rustc::internal
)]
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/a735ace01568a298cc83f5bd3e0cb7ad9fff5e82..3d16f336e2a12fccebeb2655f12bc8acb7d2b6ce)*

File: tests/lint-attributes-organization.rs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    for path in WalkDir::new(".")
+        .into_iter()
+        .flatten()
+        .map(DirEntry::into_path)
+        .filter(|p| {
+            p.extension().is_some_and(|ext| ext == "rs") && !SKIPPED_FILES.iter().any(|&skipped| p.starts_with(skipped))
+        })
+    {
+        for diag in error_for_content(&fs::read_to_string(&path).unwrap(), &attribute_regex) {
+            eprintln!("ERROR: {}: {diag}", path.to_str().unwrap());
```
Comment (ada4a) [inline tests/lint-attributes-organization.rs]:
```suggestion
            eprintln!("ERROR: {}: {diag}", path.display());
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/a735ace01568a298cc83f5bd3e0cb7ad9fff5e82..3d16f336e2a12fccebeb2655f12bc8acb7d2b6ce)*

File: tests/lint-attributes-organization.rs [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    let mut diags = vec![];
+    let mut attributes = HashSet::new();
+    for cap in attribute_regex.captures_iter(content) {
+        let attribute = &cap[1];
+        if !attributes.insert(attribute.to_owned()) {
+            diags.push(format!("duplicate top-level `#![{attribute}]` attribute"));
+        }
+        let lint_names = cap[2]
+            .lines()
+            .map(|l| l.split_once("//").map_or(l, |(b, _)| b).trim())
```
Comment (ada4a) [inline tests/lint-attributes-organization.rs:78]:
```suggestion
            // split off any trailing comments
            .map(|l| l.split_once("//").map_or(l, |(b, _)| b).trim())
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/a735ace01568a298cc83f5bd3e0cb7ad9fff5e82..3d16f336e2a12fccebeb2655f12bc8acb7d2b6ce)*

File: tests/lint-attributes-organization.rs [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        &attribute_regex,
+    );
+}
+
+fn check_errors(content: &str, expected: Vec<&str>, attribute_regex: &Regex) {
+    let errors = error_for_content(content, attribute_regex);
+    assert_eq!(errors.len(), expected.len());
+    for (error, expected) in errors.into_iter().zip(expected) {
+        assert_eq!(&error, expected);
+    }
```
Comment (ada4a) [inline tests/lint-attributes-organization.rs]: I think this might help with debugging in the future?
```suggestion
    for (i, (error, expected)) in iter::zip(errors, expected).enumerate() {
        assert_eq!(&error, expected, "the error {i} differed from the one expected");
    }
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/a735ace01568a298cc83f5bd3e0cb7ad9fff5e82..3d16f336e2a12fccebeb2655f12bc8acb7d2b6ce)*

File: tests/lint-attributes-organization.rs [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        let attribute = &cap[1];
+        if !attributes.insert(attribute.to_owned()) {
+            diags.push(format!("duplicate top-level `#![{attribute}]` attribute"));
+        }
+        let lint_names = cap[2]
+            .lines()
+            .map(|l| l.split_once("//").map_or(l, |(b, _)| b).trim())
+            .flat_map(|l| l.split_whitespace().map(|s| s.strip_suffix(',').unwrap_or(s)))
+            .collect::<Vec<_>>();
+        for (&a, &b) in lint_names.iter().zip(lint_names.iter().skip(1)) {
```
Comment (ada4a) [inline tests/lint-attributes-organization.rs]: I think you can save yourself an allocation here?
```suggestion
            ;
        for [a, b] in lint_names.array_windows() {
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/a735ace01568a298cc83f5bd3e0cb7ad9fff5e82..3d16f336e2a12fccebeb2655f12bc8acb7d2b6ce)*

File: tests/lint-attributes-organization.rs
Comment (ada4a) [inline tests/lint-attributes-organization.rs:1]: Do we want to enforce the ordering of the attribute kinds as well? warn->deny->forbid->allow->expect I think it was. We could also add it in a follow-up of course

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/a735ace01568a298cc83f5bd3e0cb7ad9fff5e82..3d16f336e2a12fccebeb2655f12bc8acb7d2b6ce)*

**Review by ada4a — CHANGES_REQUESTED** [review]

File: tests/lint-attributes-organization.rs
Comment (samueltardieu) [reply tests/lint-attributes-organization.rs:1]: Yes, that might be a future extension.

**Review by ada4a — APPROVED** [review]
Thank you:)

This probably deserves an FCP?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/b8f6009b565105262bb62d3a03230d194d28f57d..84c4bbf05d55d8e26e24356bec04a7fd08f2a343)*

**Comment by samueltardieu** [comment]
As you wish, this is not really a constraint as 99% of the test files are conformant, and a problem will be caught by anyone doing "cargo test -F internal" or by the CI with a clear error message.

**Comment by ada4a** [comment]
Oh, I meant "FCP" as in "let's ask the other team members what they think about this", not "I'm not sure the implementation is correct, let's wait it out a bit"

**Comment by samueltardieu** [comment]
That's also how I understood your "FCP". What I was saying is that it is a small change in my eyes (like maintaining the symbols in `clippy_utils::sym` sorted), but go ahead if you think we should ensure we have a consensus.

**Comment by ada4a** [comment]
My worry was that it might be deemed too pedantic.. Well, it wouldn't hurt to ask

@rustbot label I-nominated

**Comment by samueltardieu** [comment]
@ada4a Is there anything holding up the merge right now?

**Review by ada4a — APPROVED** [review]
Nope, I'm just traveling right now^^ Apologies for the delay

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17050/b8f6009b565105262bb62d3a03230d194d28f57d..84c4bbf05d55d8e26e24356bec04a7fd08f2a343)*

**Comment by samueltardieu** [comment]
GitHub is having real difficulties today
